## Kaggle 2019 ML Survey :
This Survey was conducted by Kaggle to understand what is happening with machine learning in different industries and the best ways for new data scientists to break into the field.


This survey helps us analyse behaviour patterns and marketplace trends of Data Science practitioners’ and Use of Data Science in Industry. This analysis might help Aspiring Data Scientists to understand better about trends in Data Science and achieve their goal.
This survey consists of 200+ questions , which was answered by 19717 people (not all the questions are answered by all the people). Only Few questions are answered by all of them. I have picked up few mostly answered questions. Let’s see what are they and what are people’s responses to them.

### VERY VERY IMPORTANT NOTE TO BE CONSIDERED :
The form responses have been randomized column-wise such that the responses that appear on the same row did not necessarily come from the same survey-taker.

In [1]:
!pip install plotly==4.4.1

     |████████████████████████████████| 7.3MB 2.8MB/s 
  Found existing installation: plotly 4.3.0
    Uninstalling plotly-4.3.0:
      Successfully uninstalled plotly-4.3.0


In [2]:
#import all the necessary libraries for data analysis and plotting

import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns
import plotly as py
import plotly.express as px

import os

In [3]:
print(py.__version__)
import plotly
plotly.__version__

4.4.1


'4.4.1'

In [4]:
print(os.listdir("../input"))
for i in os.listdir("../input/kaggle-survey-2019/"):
    print(i)

['kaggle-survey-2019']
multiple_choice_responses.csv
questions_only.csv
survey_schema.csv
other_text_responses.csv


Load the datasets and understand data and prepare the data (sort out questions, to which we want to analyse)

I have used the method of sorting out the ones with least amount of null answered (mostly answered questions in the dataset)

In [5]:
questions = pd.read_csv('../input/kaggle-survey-2019/questions_only.csv')
text_response = pd.read_csv('../input/kaggle-survey-2019/other_text_responses.csv')
multiple_choice = pd.read_csv('../input/kaggle-survey-2019/multiple_choice_responses.csv',low_memory=False)
survey = pd.read_csv('../input/kaggle-survey-2019/survey_schema.csv')


In [6]:
print(survey.shape)
survey.head()

(10, 35)


2019 Kaggle Machine Learning and Data Science Survey  \
0                                          Question:     
1                                  # of Respondents:     
2  Who was excluded? (0 = not excluded; 1 = exclu...     
3    If What is your age (# years)? 0-17 Is Selected     
4  If Select the title most similar to your curre...     

                            Q1  \
0  What is your age (# years)?   
1                        19717   
2                            0   
3                            0   
4                            0   

                                                 Q10  \
0  What is your current yearly compensation (appr...   
1                                              12497   
2                                                  0   
3                                                  1   
4                                                  1   

                                                 Q11  \
0  Approximately how much money have you spent on...   
1                                              12250   
2                                                  0   
3                                                  1   
4                                                  1   

                                                 Q12  \
0  Who/what are your favorite media sources that ...   
1                                              16781   
2                                                  0   
3                                                  1   
4                                                  0   

                                                 Q13  \
0  On which platforms have you begun or completed...   
1                                              16569   
2                                                  0   
3                                                  1   
4                                                  0   

                                                 Q14  \
0  What is the primary tool that you use at work ...   
1                                              15690   
2                                                  0   
3                                                  1   
4                                                  0   

                                                 Q15  \
0  How long have you been writing code to analyze...   
1                                              15627   
2                                                  0   
3                                                  1   
4                                                  0   

                                                 Q16  \
0  Which of the following integrated development ...   
1                                              14627   
2                                                  0   
3                                                  1   
4                                                  0   

                                                 Q17  ...  \
0  Which of the following hosted notebook product...  ...   
1                                              14443  ...   
2                                                  0  ...   
3                                                  1  ...   
4                                                  0  ...   

                                                 Q31  \
0  Which specific big data / analytics products d...   
1                                               7078   
2                                                  0   
3                                                  1   
4                                                  0   

                                                 Q32  \
0  Which of the following machine learning produc...   
1                                               7050   
2                                                  0   
3                                                  1   
4                                                  0   

                                                

Survey dataset contains 9 fields for 34 different (where all the 9 fields (rows) represent particular information whether such cases or scenarios are excluded or included in the survey) **"0" indicates NOT EXCLUDED whereas "1" indicates EXCLUDED**

In [7]:
print(multiple_choice.shape)
multiple_choice.head()

(19718, 246)


Time from Start to Finish (seconds)                           Q1  \
0               Duration (in seconds)  What is your age (# years)?   
1                                 510                        22-24   
2                                 423                        40-44   
3                                  83                        55-59   
4                                 391                        40-44   

                                       Q2  \
0  What is your gender? - Selected Choice   
1                                    Male   
2                                    Male   
3                                  Female   
4                                    Male   

                                       Q2_OTHER_TEXT  \
0  What is your gender? - Prefer to self-describe...   
1                                                 -1   
2                                                 -1   
3                                                 -1   
4                                                 -1   

                                          Q3  \
0  In which country do you currently reside?   
1                                     France   
2                                      India   
3                                    Germany   
4                                  Australia   

                                                  Q4  \
0  What is the highest level of formal education ...   
1                                    Master’s degree   
2                                Professional degree   
3                                Professional degree   
4                                    Master’s degree   

                                                  Q5  \
0  Select the title most similar to your current ...   
1                                  Software Engineer   
2                                  Software Engineer   
3                                                NaN   
4                                              Other   

                                       Q5_OTHER_TEXT  \
0  Select the title most similar to your current ...   
1                                                 -1   
2                                                 -1   
3                                                 -1   
4                                                  0   

                                                  Q6  \
0  What is the size of the company where you are ...   
1                               1000-9,999 employees   
2                                 > 10,000 employees   
3                                                NaN   
4                                 > 10,000 employees   

                                                  Q7  ...  \
0  Approximately how many individuals are respons...  ...   
1                                                  0  ...   
2                                                20+  ...   
3                                                NaN  ...   
4                                                20+  ...   

                                          Q34_Part_4  \
0  Which of the following relational database pro...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                          Q34_Part_5  \
0  Which of the following relational database pro...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                          Q34_Part_6  \
0  Which of the following relational database pro...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4              

In [8]:
print(text_response.shape)
text_response.head()

(19718, 28)


Q12_OTHER_TEXT  \
0  Who/what are your favorite media sources that ...   
1       "><script src=https://abels.xss.ht></script>   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                      Q13_OTHER_TEXT  \
0  On which platforms have you begun or completed...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                      Q14_OTHER_TEXT  \
0  What is the primary tool that you use at work ...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                     Q14_Part_1_TEXT  \
0  What is the primary tool that you use at work ...   
1                                             Sheets   
2                                    Microsoft Excel   
3                                                  D   
4                                    Microsoft Excel   

                                     Q14_Part_2_TEXT  \
0  What is the primary tool that you use at work ...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                     Q14_Part_3_TEXT  \
0  What is the primary tool that you use at work ...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                     Q14_Part_4_TEXT  \
0  What is the primary tool that you use at work ...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                     Q14_Part_5_TEXT  \
0  What is the primary tool that you use at work ...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                      Q16_OTHER_TEXT  \
0  Which of the following integrated development ...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                      Q17_OTHER_TEXT  ...  \
0  Which of the following hosted notebook product...  ...   
1                                                NaN  ...   
2                                                NaN  ...   
3                                                NaN  ...   
4                                                NaN  ...   

                                      Q28_OTHER_TEXT  \
0  Which of the following machine learning framew...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                      Q29_OTHER_TEXT  \
0  Which of the following cloud computing platfor...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                    

Multiple choices and Text responses both have same number of rows but different number of columns

In [9]:
#Check out all the columns in questions dataset 
for each_question in list(questions.iloc[0]):
    print(each_question)

Duration (in seconds)
What is your age (# years)?
What is your gender? - Selected Choice
In which country do you currently reside?
What is the highest level of formal education that you have attained or plan to attain within the next 2 years?
Select the title most similar to your current role (or most recent title if retired): - Selected Choice
What is the size of the company where you are employed?
Approximately how many individuals are responsible for data science workloads at your place of business?
Does your current employer incorporate machine learning methods into their business?
Select any activities that make up an important part of your role at work: (Select all that apply) - Selected Choice
What is your current yearly compensation (approximate $USD)?
Approximately how much money have you spent on machine learning and/or cloud computing products at your work in the past 5 years?
Who/what are your favorite media sources that report on data science topics? (Select all that apply

In [10]:
#Let's check the amount of missing values corresponding with mutliple choice answers
#Finding common columns between two dataframes

mcq_cols = list(multiple_choice.columns)
questions_cols = list(questions.columns)
mcq_cols_set = set(mcq_cols)
questions_cols_set = set(questions_cols)

In [11]:
list_of_common_columns = list(mcq_cols_set.intersection(questions_cols_set))

In [12]:
#Now let's check any null values for these columns within multiple choice dataframe
missing_values_percentage = multiple_choice[list_of_common_columns].isnull().sum()*100/len(multiple_choice[list_of_common_columns])
missing_values_in_numbers = multiple_choice[list_of_common_columns].isnull().sum()
missing_values_df = pd.DataFrame({'columns':multiple_choice[list_of_common_columns].columns,'percentage of missing values':missing_values_percentage,'number of missing values':missing_values_in_numbers})


In [13]:
print(missing_values_df)
print("So, We have columns with null value percentage less than 40%. At least we have 50% of information to analyse")

                                                                 columns  \
Q3                                                                    Q3   
Q14                                                                  Q14   
Q23                                                                  Q23   
Q7                                                                    Q7   
Q10                                                                  Q10   
Q6                                                                    Q6   
Q8                                                                    Q8   
Q11                                                                  Q11   
Q5                                                                    Q5   
Q15                                                                  Q15   
Q4                                                                    Q4   
Q19                                                                  Q19   
Q1          

In [14]:
multiple_choice_list_columns_set = set(multiple_choice[list_of_common_columns])
survey_columns_set = set(survey.columns)
common_columns_between_survey_and_final_columns_list = list(multiple_choice_list_columns_set.intersection(survey_columns_set))

Finally, we are going to analyze these 15 questions 

In [15]:
final_columns_list = common_columns_between_survey_and_final_columns_list
for each_question in list(multiple_choice[final_columns_list].iloc[0]):
    print(each_question)

In which country do you currently reside?
Approximately how many individuals are responsible for data science workloads at your place of business?
What is the primary tool that you use at work or school to analyze data? (Include text response) - Selected Choice
For how many years have you used machine learning methods?
What is the size of the company where you are employed?
Does your current employer incorporate machine learning methods into their business?
Approximately how much money have you spent on machine learning and/or cloud computing products at your work in the past 5 years?
Select the title most similar to your current role (or most recent title if retired): - Selected Choice
How long have you been writing code to analyze data (at work or at school)?
What is your current yearly compensation (approximate $USD)?
What programming language would you recommend an aspiring data scientist to learn first? - Selected Choice
What is the highest level of formal education that you have 

### 1. Age

In [16]:
print(multiple_choice['Q1'][0])
temporary_df = pd.DataFrame({"Age":multiple_choice['Q1'][1:].value_counts().index,"Number of people":multiple_choice['Q1'][1:].value_counts().values})


What is your age (# years)?


In [17]:
figure = px.bar(temporary_df,temporary_df['Age'],temporary_df['Number of people'],color=temporary_df['Age'],title="Age of People Who participated in the KAGGLE MACHINE LEARNING Survey")
figure.show()

In [18]:
print(survey[survey['Q1']=='1']['2019 Kaggle Machine Learning and Data Science Survey'])
print(survey['Q3'][:2])

Series([], Name: 2019 Kaggle Machine Learning and Data Science Survey, dtype: object)
0    In which country do you currently reside?
1                                        19717
Name: Q3, dtype: object




1.Most of the people who took part in this survey are in their 20's (Three among top four positions are occupied by people who are between 18-29 years of age)

2.Interestingly people who are above 50 years of age are also participating in kaggle (they are learning or working on Data Science). Even though they are less number of people, who are doing it but they are working or learning on one of the flourishing and interesting area at their later stages of career. 

## 2.Gender

In [19]:
print(multiple_choice['Q2'][0])
temporary_df = pd.DataFrame({'Gender':multiple_choice['Q2'][1:].value_counts().index,'Number of people':multiple_choice['Q2'][1:].value_counts().values})

What is your gender? - Selected Choice


In [20]:
px.bar(temporary_df,temporary_df['Gender'],temporary_df['Number of people'],color=temporary_df['Gender'],title="Gender of People Who participated in the KAGGLE MACHINE LEARNING Survey",color_discrete_sequence=px.colors.qualitative.T10)

In [21]:
px.pie(temporary_df,values = temporary_df['Number of people'],names=temporary_df['Gender'],title="Gender of People Who participated in the KAGGLE MACHINE LEARNING Survey",color_discrete_sequence=px.colors.qualitative.T10)

In [22]:
survey[survey['Q2']=='1']['2019 Kaggle Machine Learning and Data Science Survey']
survey['Q2'][:2]

0    What is your gender? - Selected Choice
1                                     19717
Name: Q2, dtype: object


1.Women participants who take part in the survey are less than 1/5 th of total participants Whereas 80% of them are Men

2.Few people around 2% of them didn't disclose their Gender and Responses of the people who are below 18years of age have not been considered

## 3.Where do you Live ?

In [23]:
print(multiple_choice['Q3'][0])
temporary_df = pd.DataFrame({'Country':multiple_choice['Q3'][1:].value_counts().index,'Number of people':multiple_choice['Q3'][1:].value_counts().values})

In which country do you currently reside?


In [24]:
px.choropleth(temporary_df['Number of people'], locations=temporary_df['Country'],
                    locationmode='country names',
                    color=temporary_df['Number of people'])

In [25]:
survey[survey['Q3']=='1']['2019 Kaggle Machine Learning and Data Science Survey']
survey['Q3'][:2]

0    In which country do you currently reside?
1                                        19717
Name: Q3, dtype: object


1.Most of them Who participate in Kaggle are from INDIA followed by USA.

2.It's quite overwhelming because we know that most of the top Product based companies has their origin in USA (from Silicon Valley). So, I thought it would top the list but for other reasons it hasn't. At the same time, it is even more pleasing to see the India on top of the list.

3.Responses of people who are less than 18 years of age have been discarded

## 4.Title

In [26]:
print(multiple_choice['Q5'][0])
temporary_df = pd.DataFrame({'Role':multiple_choice['Q5'][1:].value_counts().index,'Number of people':multiple_choice['Q5'][1:].value_counts().values})

Select the title most similar to your current role (or most recent title if retired): - Selected Choice


In [27]:
px.bar(temporary_df,temporary_df['Role'],temporary_df['Number of people'],color=temporary_df['Role'],title="Current DESIGNATION of the people who participated in Kaggle survey",color_discrete_sequence=px.colors.qualitative.Light24)

In [28]:
px.pie(temporary_df,values = temporary_df['Number of people'],names=temporary_df['Role'],title="Current DESIGNATION of the people who participated in Kaggle survey",color_discrete_sequence=px.colors.qualitative.Light24)

In [29]:
survey[survey['Q5']=='1']['2019 Kaggle Machine Learning and Data Science Survey']
survey['Q5'][:2]

0    Select the title most similar to your current ...
1                                                19107
Name: Q5, dtype: object


1.Surprisingly 21% of respondants are Students (who are still in their education (might be Under-graduation, Post-graduation or Ph.D)

2.Just with 0.4% more, Data Scientists have topped the list and almost 40% of the crowd are their Data Scientists or Students and rest of the people belong to different categories like Software Engineering, Research scientist and Others.

3.Responses of people who are less than 18 years of age have been discarded

4.Actually there are 19717 people who participated in the survey but 19107 people responded to this question

## 5. Employed Size of the company

In [30]:
print(multiple_choice['Q6'][0])
temporary_df = pd.DataFrame({'Size of the company':multiple_choice['Q6'][1:].value_counts().index,'Number of people':multiple_choice['Q6'][1:].value_counts().values})

What is the size of the company where you are employed?


In [31]:
px.bar(temporary_df,temporary_df['Size of the company'],temporary_df['Number of people'],color=temporary_df['Size of the company'],title="Size of the company of the people who are employed and participated in Kaggle survey",color_discrete_sequence=px.colors.qualitative.Set3)

In [32]:
px.pie(temporary_df,values = temporary_df['Number of people'],names=temporary_df['Size of the company'],title="Size of the company of the people who are employed and participated in Kaggle survey",color_discrete_sequence=px.colors.qualitative.Set3)

In [33]:
survey[survey['Q6']=='1']['2019 Kaggle Machine Learning and Data Science Survey']
survey['Q6'][:2]

0    What is the size of the company where you are ...
1                                                14002
Name: Q6, dtype: object


1.Majority of people are working in a company which have less than 50 people, which we typically call as STARTUPS.

2.Next highest proportion of people are working at companies who have more than 10000 employees and rest of them are working in mid-size organizations.

3.Actually there are 19717 people who participated in the survey but only 14002 people responded to this question.

4.Responses of People who are STUDENTS and Not Employed are not considered.

## 6. Individuals responsible for Data science workloads at your place of business

In [34]:
print(multiple_choice['Q7'][0])
temporary_df = pd.DataFrame({'Number of individuals responsible for Business':multiple_choice['Q7'][1:].value_counts().index,'Number of people':multiple_choice['Q7'][1:].value_counts().values})

Approximately how many individuals are responsible for data science workloads at your place of business?


In [35]:
px.bar(temporary_df,temporary_df['Number of individuals responsible for Business'],temporary_df['Number of people'],color=temporary_df['Number of individuals responsible for Business'],title="Number of individuals responsible for Data science workloads at your place of Business",color_discrete_sequence=px.colors.qualitative.Antique)

In [36]:
px.pie(temporary_df,values = temporary_df['Number of people'],names=temporary_df['Number of individuals responsible for Business'],title="Number of individuals responsible for Data science workloads at your place of Business",color_discrete_sequence=px.colors.qualitative.Antique)

In [37]:
survey[survey['Q7']=='1']['2019 Kaggle Machine Learning and Data Science Survey']
survey['Q7'][:2]

0    Approximately how many individuals are respons...
1                                                13623
Name: Q7, dtype: object


1.Majority of people responded that 20+ people are responsible for Data science workloads at their place of business

2.But almost 7000 people said that number of individuals responsible for Data science workloads at their place of business are less than 5 people.

3.Actually there are 19717 people who participated in the survey but only 13623 people responded to this question.

4.Responses of People who are STUDENTS and Not Employed are not considered for this question.

## 7. Current employer incorporating machine learning methods into their business

In [38]:
print(multiple_choice['Q8'][0])
temporary_df = pd.DataFrame({'ML methods in Business':multiple_choice['Q8'][1:].value_counts().index,'Number of people':multiple_choice['Q8'][1:].value_counts().values})

Does your current employer incorporate machine learning methods into their business?


In [39]:
px.bar(temporary_df,temporary_df['ML methods in Business'],temporary_df['Number of people'],color=temporary_df['ML methods in Business'],title="Incorporation of machine learning methods in Business",orientation='v')

In [40]:
px.pie(temporary_df,values = temporary_df['Number of people'],names=temporary_df['ML methods in Business'],color=temporary_df['ML methods in Business'],title="Incorporation of machine learning methods in Business")

In [41]:
survey[survey['Q8']=='1']['2019 Kaggle Machine Learning and Data Science Survey']
survey['Q8'][:2]

0    Does your current employer incorporate machine...
1                                                13227
Name: Q8, dtype: object


1.Majority of people responded that 20+ people are responsible for Data science workloads at their place of business

2.But almost 7000 people said that number of individuals responsible for Data science workloads at their place of business are less than 5 people.

3.Actually there are 19717 people who participated in the survey but only 13277 people responded to this question.

4.Responses of People who are STUDENTS and Not Employed are not considered for this question.

## 8.  Yearly compensation

In [42]:
print(multiple_choice['Q10'][0])
temporary_df = pd.DataFrame({'Salary range':multiple_choice['Q10'][1:].value_counts().index,'Number of people':multiple_choice['Q10'][1:].value_counts().values})

What is your current yearly compensation (approximate $USD)?


In [43]:
px.bar(temporary_df,temporary_df['Salary range'],temporary_df['Number of people'],color=temporary_df['Salary range'],title="Yearly compensation of participants",color_discrete_sequence=px.colors.qualitative.Light24)

In [44]:
px.pie(temporary_df,values = temporary_df['Number of people'],names=temporary_df['Salary range'],color=temporary_df['Salary range'],title="Yearly compensation",color_discrete_sequence=px.colors.qualitative.Light24)

In [45]:
survey[survey['Q10']=='1']['2019 Kaggle Machine Learning and Data Science Survey']
survey['Q10'][:2]

0    What is your current yearly compensation (appr...
1                                                12497
Name: Q10, dtype: object

1.Most of people more than 12% of people are working for salary range $0-$999. Even though, Data Scientist is regarded as high paid job but stats has proven this might be because of experience level and others. 

2.But, we shouldn't deny the fact that 6% of people are getting paid between $100000 - $124000.

3.Actually there are 19717 people who participated in the survey but only 12497 people responded to this question.

4.Responses of People who are STUDENTS and Not Employed are not considered for this question.


## 9. Primary tool which is used at work or school for Data Analysis

In [46]:
print(multiple_choice['Q14'][0])
temporary_df = pd.DataFrame({'Tools':multiple_choice['Q14'][1:].value_counts().index,'Number of people':multiple_choice['Q14'][1:].value_counts().values})

What is the primary tool that you use at work or school to analyze data? (Include text response) - Selected Choice


In [47]:
px.bar(temporary_df,temporary_df['Tools'],temporary_df['Number of people'],color=temporary_df['Tools'],title="Primary Tool which is used for Data Analysis",color_discrete_sequence=px.colors.qualitative.Pastel1)

In [48]:
px.pie(temporary_df,values = temporary_df['Number of people'],names=temporary_df['Tools'],color=temporary_df['Tools'],title="Primary tool which is used for Data Analysis",color_discrete_sequence=px.colors.qualitative.Pastel1)

In [49]:
survey[survey['Q14']=='1']['2019 Kaggle Machine Learning and Data Science Survey']
survey['Q14'][:2]

0    What is the primary tool that you use at work ...
1                                                15690
Name: Q14, dtype: object


1.There is no surprise as more than half of the people use Local environments such as RStudio, Jupyter-Notebook or Spyder for Data Analysis.

2.Cloud based data Softwares(Azure, AWS, GCP,..) and Advanced Statistical softwares (SPSS,SAS,..) and Business Intelligence softwares like (Tableau,Salesforce,..) are used by only 28% of people.

3.Actually there are 19717 people who participated in the survey but only 15690 people responded to this question.

4.Responses of People who are below 17 years of age are not considered for this question.

## 10. Years of experience in writing code to Analyze data

In [50]:
print(multiple_choice['Q15'][0])
temporary_df = pd.DataFrame({'yoe_in_coding_data_analysis':multiple_choice['Q15'][1:].value_counts().index,'Number of people':multiple_choice['Q15'][1:].value_counts().values})

How long have you been writing code to analyze data (at work or at school)?


In [51]:
px.bar(temporary_df,temporary_df['yoe_in_coding_data_analysis'],temporary_df['Number of people'],color=temporary_df['yoe_in_coding_data_analysis'],title="Years of experience in writing code to Analyze data",color_discrete_sequence=px.colors.qualitative.Pastel)

In [52]:
px.pie(temporary_df,values = temporary_df['Number of people'],names=temporary_df['yoe_in_coding_data_analysis'],color=temporary_df['yoe_in_coding_data_analysis'],title="Years of experience in writing code to Analyze data",color_discrete_sequence=px.colors.qualitative.Pastel)

In [53]:
survey[survey['Q15']=='1']['2019 Kaggle Machine Learning and Data Science Survey']
survey['Q15'][:2]

0    How long have you been writing code to analyze...
1                                                15627
Name: Q15, dtype: object


1.More than 50% of participants have less than 2 years of experience in writing code to analyze data.

2.21.5% of people have experience of nearly 3-5 years in writing code, which is required for data analysis.

3.Actually there are 19717 people who participated in the survey but only 15627 people responded to this question.

4.Responses of People who are below 18 years of age are not considered for this question.

## 11. Recommendation of programming language for Aspiring Data Scientists

In [54]:
print(multiple_choice['Q19'][0])
temporary_df = pd.DataFrame({'recommended_programming_language':multiple_choice['Q19'][1:].value_counts().index,'Number of people':multiple_choice['Q19'][1:].value_counts().values})

What programming language would you recommend an aspiring data scientist to learn first? - Selected Choice


In [55]:
px.bar(temporary_df,temporary_df['recommended_programming_language'],temporary_df['Number of people'],color=temporary_df['recommended_programming_language'],title="Recommended Programming Language for Aspiring Data Scientists",color_discrete_sequence=px.colors.qualitative.Safe)

In [56]:
px.pie(temporary_df,values = temporary_df['Number of people'],names=temporary_df['recommended_programming_language'],color=temporary_df['recommended_programming_language'],title="Recommended Programming Language for Aspiring Data Scientists",color_discrete_sequence=px.colors.qualitative.Safe)

In [57]:
survey[survey['Q19']=='1']['2019 Kaggle Machine Learning and Data Science Survey']
survey['Q19'][:2]

0    What programming language would you recommend ...
1                                                14377
Name: Q19, dtype: object


1.More than 78% of participants recommended to learn Python (as it is not at all surprising moreover it is also general purpose programming language).

2.Over 9% of people suggested to learn R (which is also very good language for Statistics).

3.Actually there are 19717 people who participated in the survey but only 14377 people responded to this question.

4.Responses of People who have never written code are not considered for this question.

## 13. Have you ever used a TPU 

In [58]:
print(multiple_choice['Q22'][0])
temporary_df = pd.DataFrame({'TPU Usage':multiple_choice['Q22'][1:].value_counts().index,'Number of people':multiple_choice['Q22'][1:].value_counts().values})

Have you ever used a TPU (tensor processing unit)?


In [59]:
px.bar(temporary_df,temporary_df['TPU Usage'],temporary_df['Number of people'],color=temporary_df['TPU Usage'],title="Using TPU for Data Sciences",color_discrete_sequence=px.colors.qualitative.Pastel1)

In [60]:
px.pie(temporary_df,values = temporary_df['Number of people'],names=temporary_df['TPU Usage'],color=temporary_df['TPU Usage'],title="Using TPU for Data Sciences",color_discrete_sequence=px.colors.qualitative.Pastel1)

In [61]:
survey[survey['Q22']=='1']['2019 Kaggle Machine Learning and Data Science Survey']
survey['Q22'][:2]

0    Have you ever used a TPU (tensor processing un...
1                                                14203
Name: Q22, dtype: object


1.More than 80% of participants have never used TPU for their Data Science tasks or projects.

2.Less than 17% of people used it less than 5 times and rest 3% of used it more than 6 times, few of them used it more than 25 times too amomg this 3%. 

3.Actually there are 19717 people who participated in the survey but only 14203 people responded to this question.

4.Responses of People who have never written code are not considered for this question.

## 14. Years of experience in Using Machine Learning methods

In [62]:
print(multiple_choice['Q23'][0])
temporary_df = pd.DataFrame({'experience_in_ML':multiple_choice['Q23'][1:].value_counts().index,'Number of people':multiple_choice['Q23'][1:].value_counts().values})

For how many years have you used machine learning methods?


In [63]:
px.bar(temporary_df,temporary_df['experience_in_ML'],temporary_df['Number of people'],color=temporary_df['experience_in_ML'],title="Years of experience in Using Machine Learning methods",color_discrete_sequence=px.colors.qualitative.Set3)

In [64]:
px.pie(temporary_df,values = temporary_df['Number of people'],names=temporary_df['experience_in_ML'],color=temporary_df['experience_in_ML'],title="Years of experience in Using Machine Learning methods",color_discrete_sequence=px.colors.qualitative.Set3)

In [65]:
survey[survey['Q23']=='1']['2019 Kaggle Machine Learning and Data Science Survey']
survey['Q23'][:2]

0    For how many years have you used machine learn...
1                                                14182
Name: Q23, dtype: object

1.More than 60% of participants have less than 2 years of experience in using Machine Learning algorithms, among them 35% of them have less than 1 year of experience in using ML algorithms.

2.Over 27% of people have 2-5 years of experience in ML algorithms and rest 13% of people have experience range of 5-20+ years of experience in Using Machine Learning algorithms.

3.Actually there are 19717 people who participated in the survey but only 14182 people responded to this question.

4.Responses of People who have never written code and are below 18 years of age are not considered for this question.

## 15. Highest level of formal education that you have attained or planning to attain within the next 2 years?

In [66]:
print(multiple_choice['Q4'][0])
temporary_df = pd.DataFrame({'Degree':multiple_choice['Q4'][1:].value_counts().index,'Number of people':multiple_choice['Q4'][1:].value_counts().values})

What is the highest level of formal education that you have attained or plan to attain within the next 2 years?


In [67]:
px.bar(temporary_df,temporary_df['Degree'],temporary_df['Number of people'],color=temporary_df['Degree'],title="Highest level of formal education that you have attained or planning to attain within the next 2 years?",color_discrete_sequence=px.colors.qualitative.Pastel)

In [68]:
px.pie(temporary_df,values = temporary_df['Number of people'],names=temporary_df['Degree'],color=temporary_df['Degree'],title="Highest level of formal education that you have attained or planning to attain within the next 2 years?",color_discrete_sequence=px.colors.qualitative.Pastel)

In [69]:
survey[survey['Q4']=='1']['2019 Kaggle Machine Learning and Data Science Survey']
survey['Q4'][:2]

0    What is the highest level of formal education ...
1                                                19323
Name: Q4, dtype: object


1.So, from this it is evident that most of the participants have completed thier Masters or Willing to pursue masters within next 2 years, who already are learning data sciences.

2.31% of people have Bachelor's degree or pursuing it. Followed by 13% of Ph.D candidates or aspiring Ph.D candidates who are currently active on kaggle.

3.Actually there are 19717 people who participated in the survey but only 19323 people responded to this question.

4.Responses of People who are below 18 years of age are not considered for this question.

## **Conclusion :**

People who are thinking that it is already late to learn Data Science or late to compete in Kaggle or in any data science hackathons should understand from this survey that majority of them have only less than 2 years of experience in using algorithms or programming related to Data Sciences and as per stats above, If you want to newly start learning data science then it is always recommended to start with Python. I believe that learning is always continuous process and should never stop learning. Once you stop it, you are out of the league. All the very best :)